### DISCUSSION TITLE
Add forces - NachtzusterBirdNET-Pi and BattyBirdNET-Pi

### DISCUSSION THREAD

**Discussion by rdz-oss** [Nachtzuster/BirdNET-Pi#75] [Ideas]
Hi Nachtzuster and community,

I really like your work on BirdNET-Pi. Your fork and mine (the one that detects bats also) are currently diverging - and that is ok.
Just to put it out there as an option: I wonder if we can add forces?

- You did great improvements on BirdNET-Pi and keep it alive for birders
- I added the ability to detect bats and birds (optionally)

For sake of discussion:  You could take over the bird detection part and I do bats etc on a joint project? 

Advantages:
- BattyBirdNET-Pi uses a fork of BirdNET-Analyzer allowing to use fine tuned classifiers of birds, bats or any other animals.
- We could keep the Analyzer Fork up to date with BirdNET-Analyzer
- The option to detect birds in the 'batty' version can be modified to allow the use of 48kHz microphones. That should be straight forward. Then BattyBirdNET-Pi could be used for birders who do not require/want ultra sound.
- We join forces on development of new features

Disadvanatges
- More tricky to coordinate. 
- Might not run on all older installations of BirdNET-Pi. If set back to 48kHz it should however considerably reduce the workload as compared to 384kHz.
- Some improvements would have to be reconsidered/re-implemented. But it is still manageable I believe.
- We would also have to get along well :)

What do you think? I would be open - but of course it is also fine if things stay apart.

Best regards,
Richard

**Comment by aderusha** [comment]
I think this is a great idea.  I'd love an opportunity to cast a wider net and see what else my station could pick up!

**Comment by Fish31520** [comment]
It could be an amazing idea but (I really dont know that much when it comes to coding... 😅) isn't it too hard to "mix" both repositories ?

**Comment by Nachtzuster** [comment]
> * We would also have to get along well :)

@rdz-oss That made me laugh - thanks

It would make a lot of sense to combine bird and bat detections one a certain level - a lot of 'infrastructure' is the same, and at least in theory, could be shared.

Since, as you point out both this fork and yours have been drifting away from mcguirepr89/BirdNET-Pi, I'm not sure what joining forces would technically mean: eg A crucial difference between mcguirepr89/BirdNET-Pi (and this one) and BattyBirdNET seems to be an integrated analysis 'pipeline' vs the dependence on `BirdNET-Analyzer`.
I would not be interested in maintaining a BirdNET-Analyzer fork, since the integrated seems to do a better job (eg way lower cpu usage)

Maybe the ~~models~~ classifiers you have created could be integrated into BirdNET-Pi, using BirdNET-Pi's analysis 'pipeline' as it were.

A hard requirement would always be back compatibility with existing installations.

**Reply by rdz-oss** [reply]
Hi @Nachtzuster ,

I considered the comments for a while. Technically, there can be different solutions. The key difference is indeed the dependence on a fork of BirdNET-Analyzer. 

That also has an advantage, i.e. a team in Brazil just now trained a model fine-tuned on a critically endangered bird species (Nemosia rourei) and was able to get that fine tuned classifier to run in BattyBirdNET-Pi. This setup can now be used in monitoring that endangered species and help preserve it!

A downside could be performance on less powerful Raspberry Pi systems. Judging from the general code of BirdNET-Pi there is some optimization potential in there, however. The footprint can be reduced considerably especially if a version pre-configured to run for birds on 48kHz is offered.

To make that feasibility/effort estimate, would you be so kind to list the fixes/features you made in an order of importance? I would then look at the corresponding commits you made and get a feel for the work needed. 

If it is ok for you, I might start integrating your changes onto the 'batty' fork in order of the importance you gave them. If this works well I would get back to you to clarify options.

Sound good?

Best regards,
Richard

**Reply by arne1921KF** [reply]
To add one more thought: it would be interesting to get the opinion of @kahst. I am confident they are aware of both forks, and Stefan even [commented](https://github.com/rdz-oss/BattyBirdNET-Pi/discussions/17) on adding bats to their analyser.

**Reply by kahst** [reply]
From the BirdNET side of things, I can say that we are aware of the demand for bat monitoring tools and that we'll do what we can to support all monitoring use cases - ideally through a foundation model for bioacoustics that we are currently working on. One thing that the next BirdNET version will allow is variable length inputs - which means you're not limited to three seconds anymore. This has some advantages for bats and down pitched recordings. We also just received some funding to develop a detection algorithm for North American bats, and that will certainly also benefit the BirdNET model and thus your efforts (which - by the way - I highly appreciate). So if you folks can think of anything that you think would be helpful to have from the modeling perspective - please reach out and let us know (maybe even as a feature request over at our analyzer repo).

**Reply by kahst** [reply]
Also, we just recently applied for funding to help maintain the BirdNET-Pi repo - let's hope that we can make that happen.

**Comment by Swiftnesses** [comment]
Love this idea guys!

Hold my beer while I go and install BattyBirdNET-Pi (that name!)

On another note, I finally have a docker config to install BirdNET-Pi with little / no hassle...

**Comment by fkdeboer** [comment]
Hi @rdz-oss and @Nachtzuster,
It would be fantastic to include the functionality of BattyBirdNet into the Nachtzuster fork and I would also like to contribute to that.

What is the status? How do you get along?:)

**Comment by alexbelgium** [comment]
Hi, for myself I find more value in running in parallel bats & birds detection. Of course having similar code would be ideal but already functionning systems are great :-)

For info I've made a docker for battybirdnet-pi here too that works on aarch64 and amd64 : https://github.com/alexbelgium/hassio-addons/blob/master/battybirdnet-pi/README_standalone.md

**Comment by alexbelgium** [comment]
I've started a first draft of combining both here : https://github.com/alexbelgium/BirdNET-Pi/blob/battybirdnet-pi

Logic :
- Depending on a BATS_ANALYSIS parameter, either the bats or the birds model is launched
- If birds, use the run_analysis routine that performs the model as currently ; and saves in birds.db
- If bats, uses a run_bat_analysis routine that starts a running battybirdnet-analyser server sends the file ; and saves in birds.db

The analyser seems to work

Also, I've not added the denoiser and timers it would only be if it works

**Reply by alexbelgium** [reply]
Everything is added now, but not fully tested

**Reply by fkdeboer** [reply]
Awesome! 
I will try and test it the coming weeks!

**Reply by rdz-oss** [reply]
Hi Alexandre,

this looks promising. I also intend to run / test it locally!
I am still figuring out how I feel about the integration - probably will write some comment soon on that (pros/cons, ideas for way forward). Good work ! 

BR
Richard

**Reply by alexbelgium** [reply]
Thanks, looking forward to ideas sharing. I've already put some elements here https://github.com/Nachtzuster/BirdNET-Pi/pull/336 to aggregate. In particular I'm still ambivalent about the fact of :
- forcing everyone to install battybirdnet-analyser from start (as this increases installation sizes and won't work with small devices)
- inability to run BirdNET and BattyBirdNET simultaneously at night
- timer logic

**Reply by rdz-oss** [reply]
Starting test tonight directly from your repository branch battybirenet-pi !

Immediately ran into an issue:

AttributeError: 'NoneType' object has no attribute 'strip'
Traceback (most recent call last):
  File "/usr/local/bin/birdnet_analysis.py", line 156, in <module>
   main()
  File "/usr/local/bin/birdnet_analysis.py", line 31, in main
    if conf.getint('BATS_ANALYSIS') == 1:
       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/lib/python3.11/configparser.py", line 1322, in get
    return _impl(self._name, option, raw=raw, vars=vars,
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/lib/python3.11/configparser.py", line 834, in getint
    return self._get_conv(section, option, int, raw=raw, vars=vars,
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/lib/python3.11/configparser.py", line 824, in _get_conv
   return self._get(section, conv, option, raw=raw, vars=vars,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 File "/usr/lib/python3.11/configparser.py", line 819, in _get
  return conv(self.get(section, option, **kwargs))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "BirdNET-Pi/scripts/utils/helpers.py", line 23, in get
    return value.strip('"')
          ^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'strip'

**Reply by alexbelgium** [reply]
Hi thanks, I think you need to run the update_snippets script to implement the needed variables

Indeed, here are some issues I've noted :
- Timer is a bit messed up for the moment, to confirm
- Extracted songs are not saved locally (although they are correctly recorded in birdnet.conf and the spectrogram is saved)
- Need to updated default birdnet.conf and not rely on update_snippets (your issue)

**Reply by alexbelgium** [reply]
I'm be off for one week but then I'll check on my return

**Reply by rdz-oss** [reply]
Cool! The update script got things working, thank you.

Also: I fixed another error - 'resampy' was missing! (pip3 install resampy) 
I added an info log on output of BattyBirdNET-Analyzer. Appears to be called correctly(although not bats out currently)

Looking forward to test on bats tonight!

**Reply by rdz-oss** [reply]
Ok caught my first bat with the system !  Congratulations @alexbelgium !

There are still a few things to fix (the scale is wrong , a few issues with selecting the max probability species per timeslot , replay in audible range, etc) but this will definitely happen. I will test a bit more and likely provide a pull request with some things I found. After it works well, we will have an interesting discussion on the way forward.
@pete-the-geek FYI: Alex started an integration of BattyBirdNET-Analyzer in the @Nachtzuster  BirdNET-Pi fork. We will need to discuss and align together on how to best proceed. There are several options with their respective pros and cons. 

<img width="968" alt="Screenshot 2025-04-30 at 20 47 38" src="https://github.com/user-attachments/assets/b77038c4-f525-4755-9a26-f13b66221a08" />

**Reply by alexbelgium** [reply]
Thanks very much! Spectrogram scale is already fixed actually I had a /2 factor that I removed. Glad it's proceeding the right way ; indeed its still a rough draft but if we are able to merge both it will make maintenance much easier!

**Reply by rdz-oss** [reply]
@alexbelgium I made a small change to server.py to select only the most likely id per time slot. It currently returns the entire list with probabilities for all species. How do you want to get the patch? I currently just cloned into your battybirdnet-pi branch.

```python
diff --git a/scripts/server.py b/scripts/server.py
index bedd842..c9ac6f9 100644
--- a/scripts/server.py
+++ b/scripts/server.py
@@ -439,6 +439,7 @@ def run_bats_analysis(file, host="127.0.0.1", port=7667):
             )
             resp.raise_for_status()
             data = resp.json()
+            # log.info("Checking for %s ", data)
     except Exception as e:
         log.error("Remote analysis failed for %s: %s", file.file_name, e)
         return []
@@ -454,15 +455,30 @@ def run_bats_analysis(file, host="127.0.0.1", port=7667):
             continue
         if stop < start:
             start, stop = stop, start
+            
+        max_species = ""
+        max_score = 0.0
 
         for species, score_str in entries:
             score = float(score_str)
+            # log.info('Checking for %s with confidence %s, start %s and stop %s', species, score, start, stop)
+            
             if score < min_conf or \
                (include and species not in include) or \
                (exclude and species in exclude) or \
                (PREDICTED_SPECIES_LIST and species not in PREDICTED_SPECIES_LIST):
                 continue
-            detections.append(Detection(file.file_date, start, stop, species, score))
+            
+            if score > max_score:
+                max_score = score
+                max_species = species
+        
+        if max_species not in ["", "noise","electrical","mechanical"]:        
+            detections.append(Detection(file.file_date, start, stop, max_species, max_score))
+            log.info('Detected %s with confidence %s at start %s and stop %s ', max_species, max_score, start, stop)
+        else:
+            log.info('No species detected with required confidence %s between %s and %s', min_conf, start, stop)
+            
 
     # ── 5. make every Detection safe for extract_safe()  ───────────────
     ex_len = 288000 / conf.getint('BATS_SAMPLING_RATE', fallback=256000)

```

**Reply by rdz-oss** [reply]
I also noticed more 'empty' recordings than I am used to. Perhaps the parameters for 'cutting' the recording are slightly off.

**Reply by alexbelgium** [reply]
> @alexbelgium I made a small change to server.py to select only the most likely id per time slot. It currently returns the entire list with probabilities for all species. How do you want to get the patch? I currently just cloned into your battybirdnet-pi branch.
> 
> ```python
> diff --git a/scripts/server.py b/scripts/server.py
> index bedd842..c9ac6f9 100644
> --- a/scripts/server.py
> +++ b/scripts/server.py
> @@ -439,6 +439,7 @@ def run_bats_analysis(file, host="127.0.0.1", port=7667):
>              )
>              resp.raise_for_status()
>              data = resp.json()
> +            # log.info("Checking for %s ", data)
>      except Exception as e:
>          log.error("Remote analysis failed for %s: %s", file.file_name, e)
>          return []
> @@ -454,15 +455,30 @@ def run_bats_analysis(file, host="127.0.0.1", port=7667):
>              continue
>          if stop < start:
>              start, stop = stop, start
> +            
> +        max_species = ""
> +        max_score = 0.0
>  
>          for species, score_str in entries:
>              score = float(score_str)
> +            # log.info('Checking for %s with confidence %s, start %s and stop %s', species, score, start, stop)
> +            
>              if score < min_conf or \
>                 (include and species not in include) or \
>                 (exclude and species in exclude) or \
>                 (PREDICTED_SPECIES_LIST and species not in PREDICTED_SPECIES_LIST):
>                  continue
> -            detections.append(Detection(file.file_date, start, stop, species, score))
> +            
> +            if score > max_score:
> +                max_score = score
> +                max_species = species
> +        
> +        if max_species not in ["", "noise","electrical","mechanical"]:        
> +            detections.append(Detection(file.file_date, start, stop, max_species, max_score))
> +            log.info('Detected %s with confidence %s at start %s and stop %s ', max_species, max_score, start, stop)
> +        else:
> +            log.info('No species detected with required confidence %s between %s and %s', min_conf, start, stop)
> +            
>  
>      # ── 5. make every Detection safe for extract_safe()  ───────────────
>      ex_len = 288000 / conf.getint('BATS_SAMPLING_RATE', fallback=256000)
> ```

A pull request is perfect, thanks

**Reply by alexbelgium** [reply]
> I also noticed more 'empty' recordings than I am used to. Perhaps the parameters for 'cutting' the recording are slightly off.

thanks - I'll look at it when I have some time

**Reply by alexbelgium** [reply]
@rdz-oss : a key element from my understanding is that as birdnet model works with 3 seconds 48khz mono files, you reduce the duration of the file to accomodate for the higher sample rates. As I understand, for example, instead of a 3s 48khz file you feed a shorter file at 256khz.

Is this code correct to calculate the duration according to the frequency ? So for example duration for a 256khz sample would be 3 * 48000 / 256000 = 0.5625 seconds ?

It seems your code references 288000 ; which would seem to indicate either 6 seconds mono 48khz samples or 3 seconds stero 48khz samples - however birdnet only works with mono files ; I'm not sure how to reconciliate both infos

Here is my calculation code : https://github.com/alexbelgium/BirdNET-Pi/blob/ca8fd38f07206443cb34c160188b2f2c1a23ccf7/scripts/utils/helpers.py#L19-L30

**Reply by alexbelgium** [reply]
I propose to also automatically calculate the RECORDING_LENGTH as a factor of the EXTRACTION_LENGTH : for example if EXTRACTION_LENGTH is 0.5625 seconds, then RECORDING_LENGTH could be 0.5625 * 6 = 3.375 seconds

This would avoid blanks

**Reply by alexbelgium** [reply]
Sound recording is fixed for bats in latest version - it is forced in wav to ensure that spectrogram & recorded sounds are not compressed
I've also used mono by default to reduce size

**Comment by alexbelgium** [comment]
Thanks I've implemented your code improvements for the server

**Comment by unosonic** [comment]
I've just found this thread and am wondering if there's already progress or success im merging features of  both branches? 
I'd find it great to have bat recognition in a bird/batNet like system.

**Reply by rdz-oss** [reply]
@alexbelgium  Made a fork of Nachtzusters BirdnetPi and integrated the bat detection from BattyBirdNet-Pi. I believe it might be this one: https://github.com/alexbelgium/BirdNET-Pi/tree/battybirdnet-pi  . It appears to not have been merged. I tried it and it did work. BattyBirdnet-Pi can be set to switch between bat and bird analysis. It will not however, have some of the Nachtzuster branch features.

**Reply by unosonic** [reply]
thanks @rdz-oss @alexbelgium I'll give it a try on a spare pi4. will there be further shared development on bat detection integrated into a Bird(Bat)Net-Pi? That would be so nice..

**Reply by alexbelgium** [reply]
Hi, this was a one-off mainly as I was interested in it but I don't intend to pursue dev further (as it anyways work)

**Comment by rdz-oss** [comment]
@alexbelgium  Thank you again for your efforts in integrating the BattyBirdNet-Analyzer into this fork of BirdNet-Pi. Would you kindly send me a brief  reply to battybirdnet  at mailbox.org ? I would like to ask your opinion on an idea.

**Reply by alexbelgium** [reply]
Hi, I had sent an email to you some time ago but also answers here in case you never received it due to spam filters. In any case, I am now using [birdnet-go](https://github.com/tphakala/birdnet-go) that incorporates your models in a seamless manner and is very efficient.
